lsb_release.1
.TH lsb_release 1
.SH NAME
lsb_release \- print distribution-specific information
.SH SYNOPSIS
.B lsb_release
.I "[options]"
.SH DESCRIPTION
The
.B lsb_release
command provides certain LSB (Linux Standard Base) and
distribution-specific information.
.PP
If no options are given, the \fI\-v\fP option is assumed.
.SH OPTIONS
The program follows the usual GNU command line syntax, with long
options starting with two dashes (`\-\-').
A summary of options are included below.
.TP
.B \-v, \-\-version
Show the version of the LSB against which your current installation is
compliant.  The version is expressed as a colon separated list of LSB
module descriptions.
.TP
.B \-i, \-\-id
Display the distributor's ID.
.TP
.B \-d, \-\-description
Display a description of the currently installed distribution.
.TP
.B \-r, \-\-release
Display the release number of the currently installed distribution.
.TP
.B \-c, \-\-codename
Display the code name of the currently installed distribution.
.TP
.B \-a, \-\-all
Display all of the above information.
.TP
.B \-s, \-\-short
Use the short output format for any information displayed.  This
format omits the leading header(s).
.TP
.B \-u, \-\-upstream
Display the corresponding information for the upstream release. e.g.,
the base Ubuntu release for a forked distro in stored in the file:
/etc/upstream-release/lsb-release
.TP
.B \-h, \-\-help
Show summary of options.
.SH NOTES
This is a reimplementation of the 
.B lsb_release
command provided by the Free Standards Group.  Any bugs are solely the
responsibility of the author below.

Detection of systems using a mix of packages from various distributions
or releases is something of a black art; the current heuristic tends
to assume that the installation is of the earliest distribution which
is still being used by .B apt but that heuristic is subject to error.
.SH "SEE ALSO"
lsb(8)
.SH AUTHOR
Chris Lawrence <lawrencc@debian.org>.


debian/changelog
lsb (4.0-1jolicloud2) robby; urgency=low

  * Inhibit displaying LSB-based boot/shutdown messages when "quiet"
    configured as kernel parameter, instead log to /var/log/lsb.log

 -- Adam McDaniel <adam@jolicloud.org>  Fri, 10 Dec 2010 12:36:57 -0700

lsb (4.0-1jolicloud1) robby-lucid; urgency=low

  * Add a -u,--upstream flag to lsb_release; read /etc/upstraem-
    release/lsb-release

 -- Adam McDaniel <adam@jolicloud.org>  Mon, 20 Sep 2010 15:27:07 -0600

lsb (4.0-0ubuntu8) lucid; urgency=low

  * sendmail, debian/control, debian/rules, debian/lsb-invalid-mta.dirs:
    Added new binary package "lsb-invalid-mta" containing a fake
    /usr/sbin/sendmail so that the installation of LSB packages does not
    pull in an MTA any more, which once adds awkward configuration questions
    to the installation process and second adds an often unnneeded daemon to
    the system (LP: #141641).

  * debian/control: The LSB requires the printing environment already
    since version 3.2. Made lsb-printing required by the "lsb" package
    not only suggested (LP: #386128).

 -- Till Kamppeter <till.kamppeter@gmail.com>  Mon, 22 Feb 2010 23:50:49 +0100

lsb (4.0-0ubuntu7) lucid; urgency=low

  * Remove /usr/lib/pymodules/python2.6/lsb_release.py on upgrades; this might
    have stayed behind due to using python-support in intermediate karmic
    development versions; LP: #418017.

 -- Loïc Minier <loic.minier@ubuntu.com>  Fri, 18 Dec 2009 15:34:25 +0100

lsb (4.0-0ubuntu6) lucid; urgency=low

  * lsb-base-logging-ubuntu.sh: Remove the log_to_console madness, it doesn't
    work with Plymouth anyway.

 -- Scott James Remnant <scott@ubuntu.com>  Fri, 04 Dec 2009 11:46:53 +0000

lsb (4.0-0ubuntu5) karmic; urgency=low

  * lsb-base-logging-ubuntu.sh: Honour CONSOLE environment variable and use
    it for output instead of /dev/console if it's set (LP: #440871).

 -- Colin Watson <cjwatson@ubuntu.com>  Thu, 08 Oct 2009 19:53:53 +0100

lsb (4.0-0ubuntu4) karmic; urgency=low

  * Merge 3.2-23 from Debian, fixing TERM handling in log_use_fancy_output
    (LP: #418461).
  * Use dpkg --print-architecture rather than
    --print-installation-architecture (LP: #403316).

 -- Colin Watson <cjwatson@ubuntu.com>  Thu, 10 Sep 2009 12:42:03 +0100

lsb (4.0-0ubuntu3) karmic; urgency=low

  * rebuild using python-central and DH_PYCENTRAL=include-links
    (LP: #383697)
  * debian/control: 
    - set XS-Python-Version to "all"
    - depend on python-central
  * debian/lsb-release.dirs, debian/rules:
    - install into usr/share/pyshared/

 -- Michael Vogt <michael.vogt@ubuntu.com>  Fri, 21 Aug 2009 16:44:13 +0200

lsb (4.0-0ubuntu2) karmic; urgency=low

  * debian/control: Added ghostscript-cups dependency (LP: #385606).

 -- Till Kamppeter <till.kamppeter@gmail.com>  Thu, 11 Jun 2009 23:43:49 +0200

lsb (4.0-0ubuntu1) karmic; urgency=low

  * Merge from Debian unstable, remaining changes:
    - lsb-base-logging-ubuntu.sh: change log_{success,warning,failure}_msg()
      to use usplash_write "TEXT" instead of "STATUS" to avoid unreadable
      line-wrapping when using usplash in verbose mode.
    - Don't conflict with python (>= 2.6).
    - Depend on postfix rather than exim4 as preferred mail-transport-agent
      alternative.
    - Depend on libgl1-mesa-glx rather than libgl1-mesa as preferred libgl1
      alternative.
    - Since /etc/lsb-release overrides detected information, there's no
      need to try and detect that information if lsb-release contains
      everything we need.  This saves us calling the hugely expensive
      apt-cache.
    - Add Ubuntu logging functions.

  [ Jeff Licquia ]
  * Upgrade to support LSB 4.0.

 -- Steve Langasek <steve.langasek@ubuntu.com>  Tue, 02 Jun 2009 20:20:07 +0000

lsb (3.2-23) unstable; urgency=low

  * Fix tests for $TERM in log_use_fancy_output().  (Closes: #537112, #536190)
  * Add libnss3-1d and libqt4-sql-sqlite to lsb-desktop dependencies.
    (Closes: #534277, #534452)
  * Properly trap open() failures in lsb_release.  (Closes: #521462)
  * Ignore empty arguments in /etc/lsb-release.  (Closes: #485565)
    May also fix #514041.
  * Better test for existant but unreadable PID file.  (Closes: #527890)
  * Add some punctuation to log_{warning,failure}_msg.  (Closes: #525684)
  * lsb-core depends on cups-bsd | lpr, since it is more likely that
    someone in 2009 will have CUPS installed already than want to use lpr
    instead.  (Closes: #512098)

 -- Chris Lawrence <lawrencc@debian.org>  Tue, 21 Jul 2009 04:03:41 -0500

lsb (3.2-22) unstable; urgency=low

  * Fix quoting of $PWD in start-stop-daemon call.  (Closes: #520499)

 -- Chris Lawrence <lawrencc@debian.org>  Fri, 20 Mar 2009 11:40:50 -0500

lsb (3.2-21) unstable; urgency=low

  * Provide lsb_release module for Python applications.  (Closes: #486262)
  * Change working directory to $PWD in start-stop-daemon, for better
    compatibility with LSB applications.  (Closes: #519817)
  * return 3 rather than 4 in pidofproc if a PID file is specified and
    the daemon is not found.  (Closes: #494623)
  * Improve handling of future testing versions unknown to lsb-release.
    Patch by Jan Muszynski (Closes: #517594).
  * Fix DeprecationWarning with Python 2.6; patch from Colin Watson and
    Anders Kaseorg.  (Closes: #517819)
  * Bypass guess_debian_release() if /etc/lsb-release is complete.
    Patch from Scott James Remnant.  (Closes: #511952)
  * 'time' was missing from lsb-core dependencies.  (Closes: #510488)
  * Demote lsb to a suggestion by lsb-release.  (Closes: #509611)

 -- Chris Lawrence <lawrencc@debian.org>  Thu, 19 Mar 2009 16:29:25 -0500

lsb (3.2-20ubuntu4) jaunty; urgency=low

  * Replace the deprecated Python sets module with the builtin set type,
    if it is available (python >= 2.4), to avoid spewing warnings for
    python >= 2.6.  (LP: #336436)

 -- Anders Kaseorg <andersk@mit.edu>  Mon, 02 Mar 2009 03:35:14 -0500

lsb (3.2-20ubuntu3) jaunty; urgency=low


lsb_release.py
#!/usr/bin/python

# LSB release detection module for Debian
# (C) 2005-09 Chris Lawrence <lawrencc@debian.org>

#    This package is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; version 2 dated June, 1991.

#    This package is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.

#    You should have received a copy of the GNU General Public License
#    along with this package; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
#    02111-1307, USA.

import sys
import commands
import os
import re

# XXX: Update as needed
# This should really be included in apt-cache policy output... it is already
# in the Release file...
RELEASE_CODENAME_LOOKUP = {
    '1.1' : 'buzz',
    '1.2' : 'rex',
    '1.3' : 'bo',
    '2.0' : 'hamm',
    '2.1' : 'slink',
    '2.2' : 'potato',
    '3.0' : 'woody',
    '3.1' : 'sarge',
    '4.0' : 'etch',
    '5.0' : 'lenny',
    }

TESTING_CODENAME = 'unknown.new.testing'

def lookup_codename(release, unknown=None):
    m = re.match(r'(\d+)\.(\d+)(r(\d+))?', release)
    if not m:
        return unknown

    shortrelease = '%s.%s' % m.group(1,2)
    return RELEASE_CODENAME_LOOKUP.get(shortrelease, unknown)

# LSB compliance packages... may grow eventually
PACKAGES = 'lsb-core lsb-cxx lsb-graphics lsb-desktop lsb-qt4 lsb-languages lsb-multimedia lsb-printing'

modnamere = re.compile(r'lsb-(?P<module>[a-z0-9]+)-(?P<arch>[^ ]+)(?: \(= (?P<version>[0-9.]+)\))?')

def valid_lsb_versions(version, module):
    # If a module is ever released that only appears in >= version, deal
    # with that here
    if version == '3.0':
        return ['2.0', '3.0']
    elif version == '3.1':
        if module in ('desktop', 'qt4'):
            return ['3.1']
        else:
            return ['2.0', '3.0', '3.1']
    elif version == '3.2':
        if module == 'desktop':
            return ['3.1', '3.2']
        elif module == 'qt4':
            return ['3.1']
        elif module in ('printing', 'languages', 'multimedia'):
            return ['3.2']
        elif module == 'cxx':
            return ['3.0', '3.1', '3.2']
        else:
            return ['2.0', '3.0', '3.1', '3.2']
    elif version == '4.0':
        if module == 'desktop':
            return ['3.1', '3.2', '4.0']
        elif module == 'qt4':
            return ['3.1']
        elif module in ('printing', 'languages', 'multimedia'):
            return ['3.2', '4.0']
        elif module == 'security':
            return ['4.0']
        elif module == 'cxx':
            return ['3.0', '3.1', '3.2', '4.0']
        else:
            return ['2.0', '3.0', '3.1', '3.2', '4.0']

    return [version]

try:
    set # introduced in 2.4
except NameError:
    import sets
    set = sets.Set

# This is Debian-specific at present
def check_modules_installed():
    # Find which LSB modules are installed on this system
    output = commands.getoutput("dpkg-query -f '${Version} ${Provides}\n' -W %s 2>/dev/null" % PACKAGES)
    if not output:
        return []

    modules = set()
    for line in output.split(os.linesep):
        version, provides = line.split(' ', 1)
        version = version.split('-', 1)[0]
        for pkg in provides.split(','):
            mob = modnamere.search(pkg)
            if not mob:
                continue

            mgroups = mob.groupdict()
            # If no versioned provides...
            if mgroups.get('version'):
                module = '%(module)s-%(version)s-%(arch)s' % mgroups
                modules.add(module)
            else:
                module = mgroups['module']
                for v in valid_lsb_versions(version, module):
                    mgroups['version'] = v
                    module = '%(module)s-%(version)s-%(arch)s' % mgroups
                    modules.add(module)

    modules = list(modules)
    modules.sort()
    return modules

longnames = {'v' : 'version', 'o': 'origin', 'a': 'suite',
             'c' : 'component', 'l': 'label'}

def parse_policy_line(data):
    retval = {}
    bits = data.split(',')
    for bit in bits:
        kv = bit.split('=', 1)
        if len(kv) > 1:
            k, v = kv[:2]
            if k in longnames:
                retval[longnames[k]] = v
    return retval

def parse_apt_policy():
    data = []
    
    policy = commands.getoutput('LANG=C apt-cache policy 2>/dev/null')
    for line in policy.split('\n'):
        line = line.strip()
        m = re.match(r'(\d+)', line)
        if m:
            priority = int(m.group(1))
        if line.startswith('release'):
            bits = line.split(' ', 1)
            if len(bits) > 1:
                data.append( (priority, parse_policy_line(bits[1])) )

    return data

def guess_release_from_apt(origin='Debian', component='main',
                           ignoresuites=('experimental'),
                           label='Debian'):
    releases = parse_apt_policy()

    if not releases:
        return None

    # We only care about the specified origin, component, and label
    releases = [x for x in releases if (
        x[1].get('origin', '') == origin and
        x[1].get('component', '') == component and
        x[1].get('label', '') == label)]

    # Check again to make sure we didn't wipe out all of the releases
    if not releases:
        return None
    
    releases.sort()
    releases.reverse()

    # We've sorted the list by descending priority, so the first entry should
    # be the "main" release in use on the system

    return releases[0][1]

def guess_debian_release():
    distinfo = {'ID' : 'Debian'}

    kern = os.uname()[0]
    if kern in ('Linux', 'Hurd', 'NetBSD'):
        distinfo['OS'] = 'GNU/'+kern
    elif kern == 'FreeBSD':
        distinfo['OS'] = 'GNU/k'+kern
    else:
        distinfo['OS'] = 'GNU'

    distinfo['DESCRIPTION'] = '%(ID)s %(OS)s' % distinfo

    if os.path.exists('/etc/debian_version'):
        try:
            release = open('/etc/debian_version').read().strip()
        except IOError, msg:
            print >> sys.stderr, 'Unable to open /etc/debian_version:', str(msg)
            release = 'unknown'
            
        if not release[0:1].isalpha():
            # /etc/debian_version should be numeric
            codename = lookup_codename(release, 'n/a')
            distinfo.update({ 'RELEASE' : release, 'CODENAME' : codename })
        elif release.endswith('/sid'):
            if release.rstrip('/sid').lower().isalpha() != 'testing':
                global TESTING_CODENAME
                TESTING_CODENAME = release.rstrip('/sid')
            distinfo['RELEASE'] = 'testing/unstable'
        else:
            distinfo['RELEASE'] = release

    # Only use apt information if we did not get the proper information
    # from /etc/debian_version or if we don't have a codename
    # (which will happen if /etc/debian_version does not contain a
    # number but some text like 'testing/unstable' or 'lenny/sid')
    #
    # This is slightly faster and less error prone in case the user
    # has an entry in his /etc/apt/sources.list but has not actually
    # upgraded the system.
    rinfo = guess_release_from_apt()
    if rinfo and not distinfo.get('CODENAME'):
        release = rinfo.get('version')
        if release:
            codename = lookup_codename(release, 'n/a')
        else:
            release = rinfo.get('suite', 'unstable')
            if release == 'testing':
                # Would be nice if I didn't have to hardcode this.
                codename = TESTING_CODENAME
            else:
                codename = 'sid'
        distinfo.update({ 'RELEASE' : release, 'CODENAME' : codename })

    if distinfo.get('RELEASE'):
        distinfo['DESCRIPTION'] += ' %(RELEASE)s' % distinfo
    if distinfo.get('CODENAME'):
        distinfo['DESCRIPTION'] += ' (%(CODENAME)s)' % distinfo

    return distinfo

# Whatever is guessed above can be overridden in /etc/lsb-release
def get_lsb_information(upstream=False):
    distinfo = {}
    path='/etc/lsb-release'
    if upstream:
        path='/etc/upstream-release/lsb-release'
    if os.path.exists(path):
        try:
            for line in open(path):
                line = line.strip()
                if not line:
                    continue
                # Skip invalid lines
                if not '=' in line:
                    continue
                var, arg = line.split('=', 1)
                if var.startswith('DISTRIB_'):
                    var = var[8:]
                    if arg.startswith('"') and arg.endswith('"'):
                        arg = arg[1:-1]
                    if arg: # Ignore empty arguments
                        distinfo[var] = arg
        except IOError, msg:
            print >> sys.stderr, 'Unable to open %s:' % path, str(msg)
            
    return distinfo

def get_distro_information(upstream=False):
    lsbinfo = get_lsb_information(upstream)
    # OS is only used inside guess_debian_release anyway
    for key in ('ID', 'RELEASE', 'CODENAME', 'DESCRIPTION',):
        if key not in lsbinfo:
            distinfo = guess_debian_release()
            distinfo.update(lsbinfo)
            return distinfo
    else:
        return lsbinfo

def test():
    print get_distro_information()
    print check_modules_installed()

if __name__ == '__main__':
    test()
